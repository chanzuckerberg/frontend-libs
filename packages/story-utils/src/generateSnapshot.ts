import { render, RenderResult } from '@testing-library/react';
import { getStoriesFromStoryFileExports } from './getStories';
import type { StoryData, StoryFileExports } from './getStories';
import prepareStory from './prepareStory';

type TestOptions<Args> = {
  /**
   * get the snapshot either synchronously, returning the ChildNode,
   * or asynchronously returning a promise that resolves to the ChildNode.
   */
  getSnapshot?: (
    wrapper: RenderResult,
  ) => Promise<ChildNode | null> | ChildNode | null;
  /**
   * Override the args for the story. Useful for event handlers that may be missing.
   */
  argOverrides?: Partial<Args>;
};

function getDefaultSnapshot(wrapper: RenderResult) {
  return wrapper.container.firstChild;
}

/**
 * Runs snapshot tests on all stories imported with `import * as anthology from 'storypath.stories.tsx'`
 * @param exportedStories the exports of a .stories.tsx file including the default export with additional context
 * @param options the options for the test to add necessary context specific to snapshot tests
 */
export default function generateSnapshot<
  S extends StoryFileExports<Args>,
  Args
>(
  storiesFileExports: S,
  { getSnapshot = getDefaultSnapshot, argOverrides }: TestOptions<Args> = {},
): void {
  const stories = getStoriesFromStoryFileExports(storiesFileExports);

  for (const [storyName, story] of Object.entries<StoryData>(stories)) {
    if (story.parameters?.snapshot?.disabled) continue;

    test(`${storyName} story renders snapshot`, async () => {
      const wrapper = render(prepareStory(story.storyFn, argOverrides));
      expect(await getSnapshot(wrapper)).toMatchSnapshot();
    });
  }
}
