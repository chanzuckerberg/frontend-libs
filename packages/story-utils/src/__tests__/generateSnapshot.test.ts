/**
 * @jest-environment jsdom
 */

import generateSnapshot from '../generateSnapshot';
import * as snapshotTestStoryFile from './__mock_stories__/foo.stories';

// We should see snapshots in the subdirectory
describe('auto generate snapshots for foo', () => {
  generateSnapshot(snapshotTestStoryFile);
});
