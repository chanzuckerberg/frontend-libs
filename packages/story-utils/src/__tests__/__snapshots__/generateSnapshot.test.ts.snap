// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auto generate snapshots for foo FooStory1 story renders snapshot 1`] = `foo story 1`;

exports[`auto generate snapshots for foo FooStory2 story renders snapshot 1`] = `foo story 2`;
